### ISSUE TITLE
Bug Crash Report #30 Binder.java android.os.BinderProxy.transactNative

### ISSUE THREAD

**Issue by jeremywight** [codeforgreenville/trolley-tracker-android-client#42]
@ajhodges details in http://Fabric.io, but has triggered 10 crashes on 5 users

**Comment by jeremywight** [comment]
@ajhodges this crash is now up to 458 in the past 7 days, 25% of DAU's. If you are able to take a look would be awesome:

`

Heads Up!android.os.TransactionTooLargeException; data parcel size N bytes
--
Details:This exception occurs when too much data is transferred via Parcels concurrently. The underlying Binder transaction buffer has a limited fixed size, currently 1Mb, which is shared by all transactions in progress for the process. Consequently this exception can be thrown when there are many transactions in progress even when most of the individual transactions are of moderate size.For example, `intent.putExtra(key, new LargeParcelableObject())` where the size of the `LargeParcelableObject` exceeds 1Mb will return in this exception. This is often seen when transferring bitmaps between Activities, or when saving a large amount of state between Activity configuration changes.For more information, check out the resources below.

`

**Comment by ajhodges** [comment]
https://github.com/codeforgreenville/trolley-tracker-android-client/issues/45

sorry, I commented in Meetup, I've been working on this issue but it's been a difficult one for me to solve given my lack of in-depth experience with Android. will continue to work it as time permits

**Comment by austin9350** [comment]
@ajhodges If it would be helpful to get some extra eyes on it I know a couple Android devs that might be willing to take a peek at it.

**Comment by ajhodges** [comment]
Absolutely! Definitely could use some more experienced eyes. The problem is that when I'm saving activity/fragment instance state, I'm passing an array of parcelables (trolley/route data) to the bundle, which seems to be taking up wayyy more space than it should be and violating a size limit that they've imposed in recent versions of the android API. This only surfaced when we got the new routes, which I guess it pushing it over the edge. I think I'm running into something similar to this guy and maybe I just need to switch my parcelables to serializeables:
http://nemanjakovacevic.net/blog/english/2015/03/24/yet-another-post-on-serializable-vs-parcelable/

I tried a hacky workaround that saves the instance state to disk instead to get around the size limits, but it ends up making the app hang during opening. I just don't understand how route and trolley data could be taking 1MB or even .5MB

But the app in general could probably use some refinements from people more experienced with me... when I was creating it I really tried my best to follow Android standards and keep things as maintainable as possible, but it's the only production Android app I've really written so I know it could be better. And I know there are some anti-patterns in there as well... nothing that should be impacting this issue though.

https://developer.android.com/reference/android/os/TransactionTooLargeException.html

**Comment by stampthecoder** [comment]
I looked at the thread you posted it does seem that using parcels instead
vanilla serializable objects take up more space during deserializations.
     However, more importantly when the author wrote some automation tests
to assert this behavior he found parcel/serializable objects  were being
serialized independently in the list instead of the entire serializable
list when the array was sent independent of a wrapper object.
If we are using the Intent.putExtra (assuming using same api level)
function to serialize a List of objects, this behavior can lead to bloat
and subsequent overflow issues in buffers when deserialized.

In conclusion, (I haven't looked at the code to verify) we could wrap the
 list and send it as sending as an array will serialize each route object
and will be un marshalled  in an inefficient way.

**Comment by jeremywight** [comment]
@ajhodges I asked my colleague @Jorshasaur if he might take a look at it. 

Here is the Crash Dump:
[com.codeforgvl.trolleytrackerclient_issue_30_crash_599A224000210001290AFB8AEAAF2327_d6f035f1860311e78f0656847afe9799_0_v2.txt](https://github.com/codeforgreenville/trolley-tracker-android-client/files/1243656/com.codeforgvl.trolleytrackerclient_issue_30_crash_599A224000210001290AFB8AEAAF2327_d6f035f1860311e78f0656847afe9799_0_v2.txt)

**Comment by Jorshasaur** [comment]
Howdy @jeremywight & @ajhodges!  I took a look at the code and I have a few recommendations.

1. I would first recommend dropping the use of `putExtra` in the `SplashScreen.java` in favor of a thread safe Singleton to persist the data across activities.  If `mTrolleys` ever came back with a large amount of data, it could trigger the sort of crash you're seeing. https://github.com/codeforgreenville/trolley-tracker-android-client/blob/eb853c2b60d7e9b9b9b54ab6800b515253f122e4/app/src/main/java/com/codeforgvl/trolleytrackerclient/activities/SplashScreen.java#L158-L158 

2. I would also recommend a different strategy for saving the state of the MainActivity over saving each fragment in the bundle.  You could use the Singleton I mentioned in 1, and then rebuild the view using the stored data when the activity restarts.  If it's vital to save the state if the activity closes, then I would save that state using properties in shared preferences or tables in sqllite. 

I'll put together an example when I have some more time!

**Comment by jeremywight** [comment]
Thanks Josh!

**Comment by jeremywight** [comment]
@ajhodges You going to be able to make it to CfG tonight? Would be awesome to take a look at this bug together. Had a couple of ladies who wanted to dig in last month, but we were having an issue getting the app to build in IntelliJ and would love your help!

**Comment by ajhodges** [comment]
Alas, I've been super tied up lately. This has been nagging me for months now, I promise 😞. I've kind of been hoping that someone else would be able to step in and contribute! But if not, I will get around to it eventually... and all of the other changes that need to be implemented. I haven't forgotten about yall, life has just been crazy.

They should be using Android Studio, not IntelliJ, for what it's worth!
https://developer.android.com/studio/index.html

**Comment by jeremywight** [comment]
Hey Adam! We've been trying to build Maven in IntelliJ, but will try in Android Studio direct. If you had the time maybe you could do a super basic README to set it up?

@austin9350 was able to get it built but got blocked on the Google Maps Key. Not sure if you've got access to that or not.

**Comment by jeremywight** [comment]
@stampthecoder this is the issue

**Comment by jeremywight** [comment]
@dgshoop this is the issue we're trying to solve

**Comment by ajhodges** [comment]
So this is not a Maven project it's a Gradle project, and all of the dependencies (Android SDK, etc) and everything are loaded automatically with Android Studio, so I highly recommend using that over IntelliJ for any android development.

I do have the google maps key - I can send that in an email. Thanks guys!!!

**Comment by bikeoid** [comment]
@ajhodges , what is the mechanism to keep the Google maps key out of source control?   Is there another file that the Gradle Make uses for string resources if present?

**Comment by ajhodges** [comment]
I'm not really sure what you mean by that.. I've added the two files to the .gitignore: https://github.com/codeforgreenville/trolley-tracker-android-client/blob/master/.gitignore#L34-L35

And if they aren't present, the app won't build. So you need to either get the files that I sent to Jeremy (and I believe he forwarded to you) or get your own Google API keys from google and put them in those locations.

**Comment by bikeoid** [comment]
Ok, got it now - I'll get the files from Jeremy.   Thanks!

  (Somehow it did build when I created a dummy key in strings.xml.)

**Comment by jeremywight** [comment]
Sent over to you last night @bikeoid

**Comment by bikeoid** [comment]
Ok, it appears that SharedPreferences might work - or just as suggested above, a common thread-safe data location in RAM.   This will take a bit of rework where the structures are accessed.

But possibly nothing needs to be persisted; if the app is swapped out, just reload from the server upon reopening.    I find it hard to visualize with the range of client hardware and the Android structures involved - parcel,  fragment, etc.

**Comment by bikeoid** [comment]
I have installed google-services.json and google_maps_api.xml in the locations specified in the .gitignore file and the app builds.   I am unable to run it successfully - when the MainActivity loads, it throws a maps error - 

```
11-03 10:13:39.460 18055-18055/com.codeforgvl.trolleytrackerclient D/TROLLEYTRACKER: processing trolley updates
11-03 10:13:39.500 18055-18145/com.codeforgvl.trolleytrackerclient E/Google Maps Android API: Authorization failure.  Please see https://developers.google.com/maps/documentation/android-api/start for how to correctly set up the map.
11-03 10:13:39.510 18055-18145/com.codeforgvl.trolleytrackerclient E/Google Maps Android API: In the Google Developer Console (https://console.developers.google.com)
                                                                                              Ensure that the "Google Maps Android API v2" is enabled.
                                                                                              Ensure that the following Android Key exists:
                                                                                              	API Key: [ Exact key as in google_maps_api.xml ]
                                                                                              	Android Application (<cert_fingerprint>;<package_name>): 69:82:3F:71:9D:07:D2:55:2C:06:33:47:50:A6:59:F3:6B:6E:EE:0A;com.codeforgvl.trolleytrackerclient
11-03 10:13:41.020 18055-18132/com.codeforgvl.trolleytrackerclient W/DynamiteModule: Local module descriptor class for com.google.android.gms.googlecertificates not found.
```
I have searched for solutions to related errors and studied the Google maps API directions with no success.  Is there any other step needed to run / debug the app?

**Comment by ajhodges** [comment]
Very strange. I've been able to set up my development environment on a few different computers now, and I'm pretty confident that all I've had to do was set up android studio, pull down the code, and add those two files with API keys to the project.

I can't think of what else could be wrong - but if there is an issue with using my API keys on your local machine, you can certainly generate your own keys for maps/etc in the google console and replace them in those files. Just curious, are you running the app in an emulator or do you have your android device hooked up via ADB? I do not have much experience developing with the former.

**Comment by bikeoid** [comment]
That was it - getting my own key worked.   I was misled by the Google error message.

So I'm proceeding on my physical device hooked via ADB.  I was able to trigger the crash every time by trying to store 8 routes, so now I can find and test a fix.

Not a blocker, but I couldn't get the emulator to work: it complained that OpenGL/ES wasn't available, and no combination of with/without HAXM or software emulation or host video hardware would work.  Possibly because I'm running it in a virtual machine.

**Comment by jeremywight** [comment]
Woop woop @bikeoid!!!

**Comment by bikeoid** [comment]
Good news - I think I have a proposed fix that should be ready for beta testing if it continues to look good here for the rest of the weekend.   I've fixed some of the other issues to match the IOS version behavior.

@ajhodges , should I check the source code in after the weekend for you to release?

**Comment by ajhodges** [comment]
Sure, absolutely! That's great. Thank you for taking the time to do all of this! When you are comfortable with the code go ahead and push it to master. I can test it too and then we can release when it's looking good. Can even push out to just the beta testers first if we want to be careful.

**Comment by bikeoid** [comment]
Ok, it looked good here.   It definitely should be beta tested because of the many structural changes.

The only wonky thing is route preview when trolleys are running: later on the regular stops / trolley displays were jumpy.    I'm not sure how that case worked in the original app.

**Comment by bikeoid** [comment]
Is there a beta test in progress?   The new app worked great here last weekend.

**Comment by jeremywight** [comment]
I can beta it this weekend.

**Comment by ajhodges** [comment]
I haven't had time to build and publish a beta yet, sorry guys. Had a death
in the family and have been traveling a lot. Hope to have time next week.

**Comment by ajhodges** [comment]
I was able to pull down and build the latest code. Found a couple small issues that I don't think were related to the new code. Cleaned up some of the cruft from my past attempts at fixing this bug. I also attempted to fix that route preview flickering glitch, but I don't think it's fixed yet. It must be a preexisting issue that I've never noticed before... Not a major issue either way, I'll log it and take another stab at fixing it when I have more time.

Just pushed a beta build to the app store, it'll be live soon. We can do some testing on Thursday and if all looks good promote it. Thanks for all the hard work in my absence @bikeoid !!
